src/mirage_kv_unix.ml

```diff
@@ -46,224 +71,298 @@ let split_string delimiter name =
-  Lwt.catch (fun () ->
-    let s = Lwt_unix.files_of_directory path in
-    let s = Lwt_stream.filter (fun s -> s <> "." && s <> "..") s in
-    Lwt_stream.to_list s >>= fun l ->
-    Lwt_list.fold_left_s (fun result filename ->
-        match result with
-        | Error e -> Lwt.return (Error e)
-        | Ok files ->
-          file_or_directory t (Mirage_kv.Key.add key filename) >|= function
-          | Error e -> Error e
-          | Ok kind -> Ok ((filename, kind) :: files))
-      (Ok []) l)
+  let path = resolve_filename t.base key in
+  Lwt.catch
+    (fun () ->
+      let s = Lwt_unix.files_of_directory path in
+      let s = Lwt_stream.filter (fun s -> s <> "." && s <> "..") s in
+      let* l = Lwt_stream.to_list s in
+      Lwt_list.fold_left_s
+        (fun result filename ->
+          let*? files = Lwt.return result in
+          let path = Mirage_kv.Key.add key filename in
+          let+? kind = file_or_directory t path in
+          (path, kind) :: files)
+        (Ok []) l)
     (function
-      | Unix.Unix_error (Unix.ENOENT, _, _) -> Lwt.return (Error (`Not_found key))
-      | Unix.Unix_error (Unix.ENOTDIR, _, _) -> Lwt.return (Error (`Dictionary_expected key))
-      | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
+      | Unix.Unix_error (ENOENT, _, _) -> Lwt.return (err_not_found key)
+      | Unix.Unix_error (ENOTDIR, _, _) -> Lwt.return (err_dict_expected key)
+      | Unix.Unix_error (e, _, _) -> Lwt.return (err_unix key e)
       | e -> Lwt.fail e)
 
-let batch t ?retries:_ret f = f t
-
 let digest t key =
-  let path = resolve_filename t.base (Mirage_kv.Key.to_string key) in
-  Lwt.catch (fun () -> file_or_directory t key >>= function
-    | Error e -> Lwt.return @@ Error e
-    | Ok `Value -> Lwt.return @@ Ok (Digest.file path)
-    | Ok `Dictionary -> list t key >|= function
-      | Error e -> Error e
-      | Ok entries ->
-        let kind_to_str = function
-        | `Value -> "value"
-        | `Dictionary -> "dictionary"
-        in
-        let string_entries = List.map (fun (name, kind) -> name ^ kind_to_str kind) entries in
-        Ok (Digest.string @@ String.concat "\n" string_entries))
+  let path = resolve_filename t.base key in
+  Lwt.catch
+    (fun () ->
+      let*? v = file_or_directory t key in
+      match v with
+      | `Value -> Lwt.return (Ok (Digest.file path))
+      | `Dictionary ->
+          let open Lwt_result.Syntax in
+          let+ entries = list t key in
+          let kind_to_str = function
+            | `Value -> "value"
+            | `Dictionary -> "dictionary"
+          in
+          let string_entries =
+            List.map
+              (fun (name, kind) ->
+                Mirage_kv.Key.to_string name ^ kind_to_str kind)
+              entries
+          in
+          let entries = String.concat "\n" string_entries in
+          Digest.string entries)
```

Comment:
Since mirage-kv.6.1.1 we're allowed to return `` Error (`Value_expected _) `` for dictionaries. I'm not sure how deterministic the order of the directory entries are, nor does it check the contents of the directory entries (not sure if this is required).

Reply:
I think I'm in favor of returning an error. Besides the order of the directory entries depending on the filesystem and might not be deterministic (at least not across filesystems), `digest` doesn't check the *contents* of the directory entries:
The digest of `dir` would be the same in these two examples:
```
dir/
└── subdir
    └── a
```
```
dir/
└── subdir
    ├── a
    └── b
```
I think recursing is not a good idea.